### PR TITLE
Secure Mail 2.0

### PR DESCRIPTION
**Pull request by fabrixxm**
Remove php-gpg, add openpgp-php and phpseclib.
Add template file for settings.
Add "send test" functionality.
Add messages.po

### CODE REVIEW

File: securemail/securemail.php [hunk 1]
```diff
@@ -1,92 +1,175 @@
...
+ * Name: Secure Mail
+ * Description: Send notification mail encrypted with user-defined public GPG key
+ * Version: 2.0
+ * Author: Fabio Comuni <http://kirgroup.com/profile/fabrixxm>
+ */
+
+require_once "include/Emailer.php";
+
+/* because the fraking openpgp-php is in composer, require libs in composer
+ * and then don't use autoloader to load classes... */
```
Comment (Hypolite) [inline securemail/securemail.php]: Wait, why don't we use the autoloader?

**Comment by tobiasd** [comment]
What with independent addons - not every addon for Friendica is also part of this repository -  put their dependencies into Friendica core as well? I think addons should care about their own dependencies. They are separate works building onto the platform, not necessarily part of the platform.

**Comment by Hypolite** [comment]
The potential issue with independent addons is exactly what the Composer Merge Plugin is about: by scanning the addon folder for `composer.json` files, it ensures that the base `vendor` directory and the `autoload.php` are populated with any library required by any addon.

**Comment by fabrixxm** [comment]
I was thinking something like that plugin , but I have to draft better how make it work..

back to the PR, any  comments?
Note that I used "vendor" dir because I installed the libraries with composer, but then I remove everything not strictly required, composer-related files also, as I'm including them directly..
Note also that the only code I wrote is in securemail/securemail.php (and most of the openpgp-php usage is.. inspired by a wordpress plugin...)

**Comment by Hypolite** [comment]
Aside from that janky library inclusion, the plugin seemed fine to me. I'd still really prefer the library to be available in the main Friendica repository for non-independent addons.

File: securemail/securemail.php [hunk 2]
```diff
@@ -1,92 +1,175 @@
...
-}
+<?php
+/**
+ * Name: Secure Mail
+ * Description: Send notification mail encrypted with user-defined public GPG key
+ * Version: 2.0
+ * Author: Fabio Comuni <http://kirgroup.com/profile/fabrixxm>
+ */
+
+require_once "include/Emailer.php";
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please use a consistent quoting-style throughout the file. It seems that single-quotes are more popular, I'd suggest you use them here and in the following lines.

File: securemail/securemail.php [hunk 3]
```diff
@@ -1,92 +1,175 @@
...
+
+/* because the fraking openpgp-php is in composer, require libs in composer
+ * and then don't use autoloader to load classes... */
+$path = __DIR__."/vendor/phpseclib/phpseclib/phpseclib/";
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please use a consistent format for concatenation operator spacing. It seems that the most popular style is space-wrapped, I'd suggest you apply it here and everywhere else it's missing in the file.

File: securemail/securemail.php [hunk 4]
```diff
@@ -1,92 +1,175 @@
...
+    register_hook('emailer_send_prepare',  'addon/securemail/securemail.php', 'securemail_emailer_send_prepare');
+
+    logger("installed securemail");
+}
+
+function securemail_uninstall() {
+    unregister_hook('plugin_settings', 'addon/securemail/securemail.php', 'securemail_settings');
+    unregister_hook('plugin_settings_post', 'addon/securemail/securemail.php', 'securemail_settings_post');
+
+    unregister_hook('emailer_send_prepare',  'addon/securemail/securemail.php', 'securemail_emailer_send_prepare');
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please remove the extra space after the first comma.

File: securemail/securemail.php [hunk 5]
```diff
@@ -1,92 +1,175 @@
...
+
+    logger("removed securemail");
+}
+
+
+function securemail_settings(&$a,&$s){
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas. Also please type-hint the parameters and add a leading documentation comment.

File: securemail/securemail.php [hunk 6]
```diff
@@ -1,92 +1,175 @@
...
+    if(! local_user()) {
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after `if`.

File: securemail/securemail.php [hunk 7]
```diff
@@ -1,92 +1,175 @@
...
+        return;
+    }
+
+    $enable = intval(get_pconfig(local_user(),'securemail','enable'));
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 8]
```diff
@@ -1,92 +1,175 @@
...
+    $publickey = get_pconfig(local_user(),'securemail','pkey');
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 9]
```diff
@@ -1,92 +1,175 @@
...
+    $s = replace_macros($t, array(
+        '$title' => t('"Secure Mail" Settings'),
+        '$submit' => t('Save Settings'),
+        '$test' => t('Save and send test'), //NOTE: update also in 'post'
+        '$enable' => array('securemail-enable', t('Enable Secure Mail'), $enable, ""),
+        '$publickey' => array('securemail-pkey', t('Public key'), $publickey, t("Your public PGP key, ascii armored format"), "rows='10'")
+    ));
+
+
+
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please remove extraneous new lines

File: securemail/securemail.php [hunk 10]
```diff
@@ -1,92 +1,175 @@
...
+    return;
+}
+function securemail_settings_post(&$a, &$b){
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a new line after the last function, type-hint to arguments and a leading documentation comment block.

File: securemail/securemail.php [hunk 11]
```diff
@@ -1,92 +1,175 @@
...
+
+    if(! local_user()) {
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after `if`.

File: securemail/securemail.php [hunk 12]
```diff
@@ -1,92 +1,175 @@
...
+        return;
+    }
+
+    if($_POST['securemail-submit']) {
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after `if`.

File: securemail/securemail.php [hunk 13]
```diff
@@ -1,92 +1,175 @@
...
+        set_pconfig(local_user(),'securemail','pkey',trim($_POST['securemail-pkey']));
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 14]
```diff
@@ -1,92 +1,175 @@
...
+        $enable = ((x($_POST,'securemail-enable')) ? 1 : 0);
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 15]
```diff
@@ -1,92 +1,175 @@
...
+        set_pconfig(local_user(),'securemail','enable', $enable);
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 16]
```diff
@@ -1,92 +1,175 @@
...
+        info( t('Secure Mail Settings saved.') . EOL);
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please remove the space after the opening parenthesis.

File: securemail/securemail.php [hunk 17]
```diff
@@ -1,92 +1,175 @@
...
+
+        if ($_POST['securemail-submit'] == t('Save and send test')) {
+            $sitename = $a->config['sitename'];
+
+            $hostname = $a->get_hostname();
+            if (strpos($hostname, ':')){
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space before the opening brace.

File: securemail/securemail.php [hunk 18]
```diff
@@ -1,92 +1,175 @@
...
+                $hostname = substr($hostname, 0, strpos($hostname, ':'));
+            }
+
+            $sender_email = $a->config['sender_email'];
+            if (empty($sender_email)){
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a before opening braces.

File: securemail/securemail.php [hunk 19]
```diff
@@ -1,92 +1,175 @@
...
+                $sender_email = 'noreply@'.$hostname;
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please wrap concatenation operators with spaces.

File: securemail/securemail.php [hunk 20]
```diff
@@ -1,92 +1,175 @@
...
+            }
+
+            $subject = "Friendica - Secure Mail - Test";
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please use single quotes.

File: securemail/securemail.php [hunk 21]
```diff
@@ -1,92 +1,175 @@
...
+                'fromName' => $sitename,
+                'fromEmail' => $sender_email,
+                'toEmail' => $a->user['email'],
+                'messageSubject' => $subject,
+                'htmlVersion' => "<p>{$message}</p>",
+                'textVersion' => $message,
+            );
+
+            // enable addon for test
+            set_pconfig(local_user(),'securemail','enable', 1);
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 22]
```diff
@@ -1,92 +1,175 @@
...
+
+            $res = Emailer::send($params);
+
+            // revert to saved value
+            set_pconfig(local_user(),'securemail','enable', $enable);
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 23]
```diff
@@ -1,92 +1,175 @@
...
+
+            if ($res) {
+                info( t("Test email sent") . EOL);
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please remove the space after opening parentheses.

File: securemail/securemail.php [hunk 24]
```diff
@@ -1,92 +1,175 @@
...
+            } else {
+                notice( t("There was an error sending the test email") .EOL);
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please remove the space after the opening parentheses and add a space after the concatenation operator.

File: securemail/securemail.php [hunk 25]
```diff
@@ -1,92 +1,175 @@
...
+            }
+        }
+    }
+}
+
+function securemail_emailer_send_prepare(&$a, &$b) {
+    if (!x($b,'uid')) {
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 26]
```diff
@@ -1,92 +1,175 @@
...
+            if ($res) {
+                info( t("Test email sent") . EOL);
+            } else {
+                notice( t("There was an error sending the test email") .EOL);
+            }
+        }
+    }
+}
+
+function securemail_emailer_send_prepare(&$a, &$b) {
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add type hint to parameters and a leading documentation comment block.

File: securemail/securemail.php [hunk 27]
```diff
@@ -1,92 +1,175 @@
...
+    if (!x($b,'uid')) {
+        return;
+    }
+
+    $uid = $b['uid'];
+
+    $enable_checked = get_pconfig($uid,'securemail','enable');
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 28]
```diff
@@ -1,92 +1,175 @@
...
+    if (!$enable_checked) {
+        return;
+    }
+
+    $public_key_ascii = get_pconfig($uid,'securemail','pkey');
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php (hunk 5 above)
Comment (fabrixxm) [inline securemail/securemail.php]: docs for those functions should links hooks docs, as are all hooks callbacks. Is there a way to do that?

Comment (fabrixxm) [inline securemail/securemail.php]: And the `$a` is the `App` class instance. Could be useful to link to it's docs, too...

File: securemail/securemail.php [hunk 29]
```diff
@@ -6,88 +6,104 @@
...
+ * @brief Build user settings form
+ *
+ * @link https://github.com/friendica/friendica/blob/develop/doc/Plugins.md#plugin_settings 'plugin_settings' hook
+ *
+ * @param App $a App instance
+ * @param string $s output html
+ *
+ * @see App
+ */
+function securemail_settings(App &$a, string &$s){
```
Comment (Hypolite) [inline securemail/securemail.php]: I'm sorry the `string` type hint is only available from PHP 7 onward, and we still have to support PHP5.

File: securemail/securemail.php [hunk 30]
```diff
@@ -100,35 +116,45 @@ function securemail_settings_post(&$a, &$b){
...
+ *
+ * @link https://github.com/friendica/friendica/blob/develop/doc/Plugins.md#emailer_send_prepare 'emailer_send_prepare' hook
+ *
+ * @param App $a App instance
+ * @param array $b hook data
+ *
+ * @see App
+ */
+function securemail_emailer_send_prepare(App &$a, array &$b) {
     if (!x($b,'uid')) {
```
Comment (Hypolite) [inline securemail/securemail.php]: Standards: Please add a space after commas.

File: securemail/securemail.php [hunk 31]
```diff
@@ -141,7 +167,12 @@ function securemail_emailer_send_prepare(&$a, &$b) {
         'filename' => 'encrypted.gpg'
     ));
     $encrypted = OpenPGP_Crypt_Symmetric::encrypt($key, new OpenPGP_Message(array($data)));
-    $armored_encrypted = wordwrap(OpenPGP::enarmor($encrypted->to_bytes(), "PGP MESSAGE"), 64, "\n", true);
+    $armored_encrypted = wordwrap(
+        OpenPGP::enarmor($encrypted->to_bytes(), 'PGP MESSAGE'),
+        64,
+        '\n',
```
Comment (Hypolite) [inline securemail/securemail.php]: I believe the intended behavior was to add an actual end of line character, not the literal `\n` characters. Usage of `PHP_EOL` is suggested.

File: securemail/securemail.php [hunk 32]
```diff
@@ -6,88 +6,104 @@
...
-            if (empty($sender_email)){
-                $sender_email = 'noreply@'.$hostname;
+            if (empty($sender_email)) {
+                $sender_email = 'noreply@' . $hostname;
             }
 
-            $subject = "Friendica - Secure Mail - Test";
-            $message = "This is a test message from your Friendica Secure Mail addon.\n\nBye!";
+            $subject = 'Friendica - Secure Mail - Test';
+            $message = 'This is a test message from your Friendica Secure Mail addon.\n\nBye!';
```
Comment (Hypolite) [inline securemail/securemail.php]: The two line feeds at the end are lost with the change of quotation style, please use `PHP_EOL` instead.

**Pull request merged**